Add structure to solve optimization problems

S-Dafarra opened the pull request:
An optimization problem can be set as follows
```py
    @dataclasses.dataclass
    class SwitchVar(OptimizationObject):
        x: StorageType = default_storage_field(ContinuousVariable)
        y: StorageType = default_storage_field(ContinuousVariable)

        def __post_init__(self):
            self.x = np.zeros(1)
            self.y = np.zeros(1)

    a = 10
    new_problem = OptimizationProblem()
    new_variables = new_problem.generate_optimization_objects(SwitchVar())
    new_problem.add_expression(
        ExpressionType.minimize, a * new_variables.y * new_variables.y
    )
    new_problem.add_expression(
        ExpressionType.subject_to, new_variables.x + new_variables.y == a - 1
    ) 
    new_problem.add_expression(ExpressionType.minimize, new_variables.x == 5)
    output, cost_value = new_problem.solver().solve()
```

Note the line 
```py
new_problem.add_expression(ExpressionType.minimize, new_variables.x == 5)
```
Equalities can be cast into a cost. This might simplify the setting of tracking tasks.

traversaro commented on test/test_optimization_problem.py:
```diff
@@ -0,0 +1,256 @@
+        c.append(20.0 * np.random.rand(3) - 10.0)
+        initial_guess[j].parameter = c[j]
+
+    problem.add_expression(
+        mode=ExpressionType.minimize,
+        expression=(
+            a[j][k] * cs.power(var[j].composite.variable[k], 2)
+            + b[j][k] * var[j].composite.variable[k]
+            for j in range(len(initial_guess))
+            for k in range(0, 3)
+        ),
+    )
+
+    problem.add_expression(
+        mode=ExpressionType.subject_to,
+        expression=(  # noqa
+            var[j].composite.variable[k] >= c[j][k]
+            for j in range(len(initial_guess))
+            for k in range(3)
+        ),
+    )
```
I am mostly looked at how user code looks, and I was a bit confused by the `add_expression` method. First of all, I was a bit confused by `add_expression` meant at a first glance, perhaps separated methods for cost such as `add_constraint` and `add_cost` make more sense. 
More specifically, it is not clear to me what happens when calls `problem.add_expression(mode=ExpressionType.minimize` . Is this substituting the existing cost, or being added to it? If the second case is true, probably a good name for this function could be `add_to_cost`. If we have functions with just one argument, we could also drop the named argument `expression=`, and increse the information density reducing the noise, i.e. :

~~~py
    problem.add_to_cost(
        (
            a[j][k] * cs.power(var[j].composite.variable[k], 2)
            + b[j][k] * var[j].composite.variable[k]
            for j in range(len(initial_guess))
            for k in range(0, 3)
        ),
    )

    problem.add_constraint(
        (  # noqa
            var[j].composite.variable[k] >= c[j][k]
            for j in range(len(initial_guess))
            for k in range(3)
        ),
    )
~~~

traversaro replied:
For example, Drake uses `add_cost` and `add_constraint`, see https://github.com/RobotLocomotion/drake/blob/v1.14.0/tutorials/mathematical_program.ipynb .

S-Dafarra replied:
The idea was to allow defining an expression, and then deciding at runtime what you want to do with it. For the time being the options are ``minimize``, ``subject_to``, and ``skip``. This is an idea I borrowed from the `blf`` ik where each task has a given priority that can be set from a configuration file. The idea is to easily allow moving a cost portion to constraints and vice-versa.

> Is this substituting the existing cost, or being added to it? 

At the moment every call adds a new cost (or constraint). There is no coming back.

Regarding the named argument, this is something of ``python``, it is not mandatory, but usually suggested.

Anyhow, if you consider it confusing, I can also add two specific methods ``add_constraint`` and ``add_cost`` that can easily coexist with the ``add_expression`` versions for people that can't decide until the last moment

S-Dafarra replied:
Option B. That's why it is called ``add_expression`` and not ``set_expression``

traversaro replied:
> Anyhow, if you consider it confusing, I can also add two specific methods add_constraint and add_cost that can easily coexist with the add_expression versions for people that can't decide until the last moment

I think this may be convenient. Just for me to understand, can you make an example of a case in which you first add an expression as `minimize`, and then you convert it to `subject_to`? It is not clear to me how this can work as there is no expression identifier.

S-Dafarra replied:
I am assuming you wanted to set ``ExpressionType.minimize``, because otherwise you are adding two constraints

traversaro replied:
> I am assuming you wanted to set `ExpressionType.minimize`, because otherwise you are adding two constraints

Yes, sorry, I fixed it now.

S-Dafarra replied:
> > Anyhow, if you consider it confusing, I can also add two specific methods add_constraint and add_cost that can easily coexist with the add_expression versions for people that can't decide until the last moment
> 
> I think this may be convenient. Just for me to understand, can you make an example of a case in which you first add an expression as `minimize`, and then you convert it to `subject_to`? It is not clear to me how this can work as there is no expression identifier.

You might set the mode from a configuration file, thus enforcing or relaxing a task without editing the code.

traversaro replied:
> > > Anyhow, if you consider it confusing, I can also add two specific methods add_constraint and add_cost that can easily coexist with the add_expression versions for people that can't decide until the last moment
> > 
> > 
> > I think this may be convenient. Just for me to understand, can you make an example of a case in which you first add an expression as `minimize`, and then you convert it to `subject_to`? It is not clear to me how this can work as there is no expression identifier.
> 
> You might set the mode from a configuration file, thus enforcing or relaxing a task without editing the code.

Ack, so basically you save an if. My intuition is that by adding also `add_cost` and `add_constraint` on top of `add_expression` you get the best of both worlds.

S-Dafarra replied:
Done in https://github.com/ami-iit/hippopt/pull/4/commits/109859cd371d860eb18993ae3bedd249d5e4ca2c

traversaro reviewed (approved):
Ok, some suggestions on the API but non blocking

S-Dafarra commented:
> Another suggestion would be to add named arguments to have a scaling when passing a costraint and cast it into a cost.

Done in https://github.com/ami-iit/hippopt/pull/4/commits/109859cd371d860eb18993ae3bedd249d5e4ca2c

diegoferigo commented on src/hippopt/base/optimization_solver.py:
```diff
@@ -0,0 +1,48 @@
+    @abc.abstractmethod
+    def generate_optimization_objects(
+        self, input_structure: TOptimizationObject | List[TOptimizationObject], **kwargs
+    ) -> TOptimizationObject | List[TOptimizationObject]:
```
While it's intuitive returning a single `TOptimizationObject` in case only one is generated, in practice it might be better to always return a list, with just one element in this case. Otherwise, downstream users that want to write generic code need to add logic to handle the two cases. If they expect just a single object, the can always hardcode `out[0]`.

diegoferigo commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+    ) -> TOptimizationObject | List[TOptimizationObject]:
+        return self._variables
+
+    def set_initial_guess(
+        self, initial_guess: TOptimizationObject | List[TOptimizationObject]
+    ):
```
Just to understand, this only sets the attribute variables and not the paramenters, right?

S-Dafarra replied:
No, it sets both. You can skip the initialization of a given variable if you use ``None``

diegoferigo replied:
Ok

diegoferigo reviewed (approved):
Nice @S-Dafarra :rocket: I like the direction. I left (too many) comments on this PR, feel free to include here what you think useful and -hopefully- address the nits/suggestions/python syntax enhancements in next ones. Approving regardless so you can move forward.

diegoferigo commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+        default=None
+    )
+
+    def __post_init__(self, problem_type: str) -> None:
+        self._solver = cs.Opti(problem_type)
+        self._solver.solver(
+            self._inner_solver, self._options_plugin, self._options_solver
+        )
```
I noticed later that there is a method to change the solver options. I'd prefer anyway to freeze all the configuration at construction time as suggested here.

S-Dafarra replied:
I would prefer not. Back in time, I needed to change the verbosity of ``ipopt`` between different runs. I would like to do that without reconstructing the whole problem.

diegoferigo replied:
What about both of them? Using an approach like the one I proposed for those that want to extend the default options during construction, and then also allow in a later moment to change the options through the method.

S-Dafarra replied:
That's reasonable! I have done it in https://github.com/ami-iit/hippopt/pull/4/commits/57e4a17955809a4570be0090d9e38fe8b301f3bf

S-Dafarra commented on src/hippopt/base/optimization_solver.py:
```diff
@@ -0,0 +1,48 @@
+    @abc.abstractmethod
+    def generate_optimization_objects(
+        self, input_structure: TOptimizationObject | List[TOptimizationObject], **kwargs
+    ) -> TOptimizationObject | List[TOptimizationObject]:
```
That depends on the input. Maybe we can enforce it with the ``Generic`` thingy you mentioned yesterday?

diegoferigo replied:
Yep, this seems a good use-case for that

S-Dafarra replied:
I have checked how to do it and I would like to implement it. Nonetheless, this would badly conflict with some code I started developing before receiving the review (https://github.com/ami-iit/hippopt/tree/ocStructure). I will tackle this in the next PR.

Reference issue: https://github.com/ami-iit/hippopt/issues/5

S-Dafarra commented on src/hippopt/base/optimization_solver.py:
```diff
@@ -0,0 +1,48 @@
+        pass
+
+    @abc.abstractmethod
+    def cost_function(self) -> cs.MX:
```
That was meant to return the full cost function, but yes, I can hide it inside the ``Opti`` class only.

S-Dafarra replied:
Done in https://github.com/ami-iit/hippopt/pull/4/commits/06e01fa5ca48237649830e5f18e4c12037064522

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+class OptiSolver(OptimizationSolver):
+    DefaultSolverType: ClassVar[str] = "ipopt"
+    _inner_solver: str = dataclasses.field(default=DefaultSolverType)
+    problem_type: dataclasses.InitVar[str] = dataclasses.field(default="nlp")
```
I wanted to keep a clean interface toward ``opti``. If ``opti`` adds a new solver type, we could avoid to edit the code

S-Dafarra replied:
I agree, although I would blame ``Opti`` for that :grin:

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+                            + " has number of dimensions greater than 2."
+                        )
+                    if value.ndim < 2:
+                        value = np.expand_dims(value, axis=1)
```
I used it at first, but it expands in the rows direction. E.g., a numpy array of size 3 was becoming of size (1,3), while I needed the opposite.

diegoferigo replied:
Ahn now I understand what you wanted to achieve. It looks good.

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+        self._opti_solution = self._solver.solve()
+        self._output_cost = self._opti_solution.value(self._cost)
+        self._output_solution = self._generate_solution_output(self._variables)
+        return self._output_solution, self._output_cost
```
We return a numeric solution

diegoferigo replied:
Numeric solution could either be a cs.DM or a np.ndarray, sorry if this was not clear :)

S-Dafarra replied:
Oh, I did not think about ``cs.DM``. The line that fills the values is https://github.com/ami-iit/hippopt/blob/109859cd371d860eb18993ae3bedd249d5e4ca2c/src/hippopt/base/opti_solver.py#L152

I am not sure how to make sure that the output is a ndarray. I wonder of this typing has effect: https://github.com/ami-iit/hippopt/blob/109859cd371d860eb18993ae3bedd249d5e4ca2c/src/hippopt/base/optimization_object.py#L10

I checked with the debugger and it seems a ``ndarray``: 
![image](https://user-images.githubusercontent.com/18591940/230344878-d4a1d1b4-fc93-4c7d-8df9-9f046439ad86.png)

S-Dafarra replied:
Added explicit casting in https://github.com/ami-iit/hippopt/pull/4/commits/3310a860b02c8f25a80e42812ef6238ba1bfdd0f

S-Dafarra commented on test/test_optimization_problem.py:
```diff
@@ -0,0 +1,254 @@
+    assert problem.solver().get_cost_value() == pytest.approx(expected_cost)
+
+
+@dataclasses.dataclass
+class TestVarAndPar(OptimizationObject):
+    composite: TestVar = dataclasses.field(default_factory=TestVar)
+    parameter: StorageType = default_storage_field(Parameter)
```
I did not get it :blush:

S-Dafarra replied:
I have done it in https://github.com/ami-iit/hippopt/pull/4/commits/34ba2dd1f1c28c03de676d35c630f65cf3a1dc65. Let me know if you meant something different.

S-Dafarra commented on src/hippopt/base/optimization_problem.py:
```diff
@@ -0,0 +1,103 @@
+            input_structure=input_structure
+        )
+
+    def add_cost(
+        self,
+        expression: cs.MX | Generator[cs.MX, None, None],
+        scaling: float | cs.MX = 1.0,
+    ):
```
The problem would be that it is difficult to associate an entry of the list with a meaningful tag. In other words, if I have a method returning ``costs = list[cs.MX]``, how do I understand what is ``cost[5]``? For the time being, I leave this on the user.

S-Dafarra replied:
Also, this would complexify the rebase of a branch I have been working on. I have spun off an issue https://github.com/ami-iit/hippopt/issues/6 and I will be dealing with this on the next PR

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+            or list_of_optimization_objects
+        )
+
+        output = copy.deepcopy(input_structure)
+        for i in range(len(output)):
+            output[i] = self.generate_optimization_objects(output[i])
```
I tried, but it does not work. They do different things. The one you propose does not create a list if the input is a list

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+    def generate_optimization_objects(
+        self, input_structure: TOptimizationObject | List[TOptimizationObject], **kwargs
+    ) -> TOptimizationObject | List[TOptimizationObject]:
+        if isinstance(input_structure, OptimizationObject):
+            return self._generate_objects_from_instance(input_structure=input_structure)
+        return self._generate_objects_from_list(input_structure=input_structure)
```
I tried, but it does not work because it returns a function, not an object

diegoferigo replied:
The first () return a function selecting one among the two options, the second () passes to it the argument, this should work fine

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+                and field.metadata[OptimizationObject.StorageTypeField]
+                == ContinuousVariable.StorageType
+            ):
+                value = output.__dict__[field.name]
```
Done in https://github.com/ami-iit/hippopt/pull/4/commits/a93f2bb0eebfea1e33c89ecc58d2d66d39291437

S-Dafarra commented on src/hippopt/base/optimization_solver.py:
```diff
@@ -0,0 +1,48 @@
+        pass
+
+    @abc.abstractmethod
+    def get_solution(self) -> TOptimizationObject | List[TOptimizationObject] | None:
```
Removed in https://github.com/ami-iit/hippopt/pull/4/commits/06e01fa5ca48237649830e5f18e4c12037064522

S-Dafarra commented on src/hippopt/base/optimization_solver.py:
```diff
@@ -0,0 +1,48 @@
+        pass
+
+    @abc.abstractmethod
+    def get_cost_value(self) -> float | None:
```
Done in https://github.com/ami-iit/hippopt/pull/4/commits/26ccb77a403f81163c9bd2032d20f62b388f8b28 and https://github.com/ami-iit/hippopt/pull/4/commits/dfdb8116f854be4df6338f8f7e925cfb42ed207c

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+
+    def _generate_objects_from_instance(
+        self, input_structure: TOptimizationObject
+    ) -> TOptimizationObject:
```
As https://github.com/ami-iit/hippopt/pull/4#discussion_r1159629096

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+    def _generate_solution_output(
+        self, variables: TOptimizationObject | List[TOptimizationObject]
+    ) -> TOptimizationObject | List[TOptimizationObject]:
+        output = copy.deepcopy(variables)
```
As https://github.com/ami-iit/hippopt/pull/4#discussion_r1159629096

S-Dafarra commented on src/hippopt/base/opti_solver.py:
```diff
@@ -0,0 +1,400 @@
+            if (
+                has_storage_field
+                and field.metadata[OptimizationObject.StorageTypeField]
+                == ContinuousVariable.StorageType
```
Done in https://github.com/ami-iit/hippopt/pull/4/commits/875bd1a2f3a41dad61bb56372093804de7ea7fc1

S-Dafarra commented:
@diegoferigo I should have addressed your comments (except https://github.com/ami-iit/hippopt/pull/4#discussion_r1154222309 where I am not sure how to continue).

I set the comments to resolved quite arbitrarily, so feel free to "Unresolve" the points that need some further discussion.

Let me know how you would like to proceed :wink:

S-Dafarra commented:
I will proceed in merging this. Other open points will be addressed in future PRs.
